src/lib.rs

```diff
@@ -97,18 +97,27 @@ pub fn repaint_stack() {
 /// In other words: shows the worst case free stack space since [repaint_stack] was last called.
 ///
 /// Runs in *O(n)* where *n* is the size of the stack.
-#[allow(clippy::manual_slice_size_calculation)]
 pub fn stack_painted() -> u32 {
-    // Safety: we should be able to read anywhere on the stack using this.
-    let slice = unsafe {
-        &*core::ptr::slice_from_raw_parts(stack().end, current_stack_free() as usize / 4)
+    let res: *const u32;
+    unsafe {
+        asm!(
+            "0:",
+            "cmp sp, {ptr}",
+            "bls 1f",
+            "ldr {value}, [{ptr}]",
+            "cmp {value}, {paint}",
+            "bne 1f",
+            "add {ptr}, #4",
+            "b 0b",
+            "1:",
+            ptr = inout(reg) stack().end => res,
+            value = out(reg) _,
+            paint = in(reg) STACK_PAINT_VALUE,
+            options(nostack, readonly)
```

Comment:
We *maybe* don't want to claim `nostack`, I asked on zulip [#t-opsem > Stack Painting, and reading in inline ASM @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/136281-t-opsem/topic/Stack.20Painting.2C.20and.20reading.20in.20inline.20ASM/near/545228787) for the opsem call on this.

Reply:
Our best guess is this is fine, actually, OK to resolve this unless opsem yells about it later.